src/ringbuffer.rs

```diff
@@ -0,0 +1,369 @@
+
+/// Transient backing data that is the backbone of the trait object.
+pub struct RingBufferTransient<I, B, M, T>
+where
+	I: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	start: Index,
+	end: Index,
+	_t: PhantomData<(I, B, M, T)>,
+}
+
+impl<I, B, M, T> RingBufferTransient<I, B, M, T>
+where
+	I: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	/// Create a new RingBufferTransient that backs the ringbuffer implementation.
+	///
+	/// Initializes itself from the Bounds.
+	pub fn new() -> RingBufferTransient<I, B, M, T> {
+		let (start, end) = <<T as RingBufferTrait<I>>::Bounds>::get();
+		RingBufferTransient {
+			start,
+			end,
+			_t: PhantomData,
+		}
+	}
+}
+
+impl<I, B, M, T> Drop for RingBufferTransient<I, B, M, T>
+where
+	I: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	fn drop(&mut self) {
+		<Self as RingBufferTrait<I>>::commit(self);
+	}
+}
+
+/// Trait object presenting the ringbuffer interface.
+pub trait RingBufferTrait<I>
+where
+	I: Codec + EncodeLike,
+{
+	type Bounds: StorageValue<(Index, Index)>;
+	type Map: StorageMap<Index, I>;
+
+	/// Store all changes made in the underlying storage.
+	///
+	/// Data is not guaranteed to be consistent before this call.
+	/// Will be called by `drop`.
+	fn commit(&self);
+
+	/// Push an item onto the end of the queue.
+	fn push(&mut self, i: I);
+	/// Push an item onto the front of the queue.
+	fn push_front(&mut self, i: I);
+
+	/// Pop an item from the start of the queue.
+	///
+	/// Returns `None` if the queue is empty.
+	fn pop(&mut self) -> Option<I>;
+}
+
+/// Ringbuffer implementation based on `RingBufferTransient`
+impl<I, B, M, T> RingBufferTrait<I> for RingBufferTransient<I, B, M, T>
+where
+	I: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	type Bounds = B;
+	type Map = M;
+
+	/// Commit the (potentially) changed bounds to storage.
+	fn commit(&self) {
+		Self::Bounds::put((self.start, self.end));
+	}
+
+	/// Push an item onto the end of the queue.
+	///
+	/// Will insert the new item, but will not update the bounds in storage.
+	fn push(&mut self, item: I) {
+		Self::Map::insert(self.end, item);
+		// this will intentionally overflow and wrap around when bonds_end
+		// reaches `Index::max_value` because we want a ringbuffer.
+		let next_index = self.end.wrapping_add(1);
+		if next_index == self.start {
+			// overwrite the oldest item in the FIFO ringbuffer
+			self.start = self.start.wrapping_add(1);
+			self.end = next_index;
+		} else {
+			self.end = next_index;
```

Comment:
nitpick but do we need the else?
```rust
if next_index == self.start {
    self.start = self.start.wrapping_add(1);
}
self.end = next_index;
```
Or do you find it more readable the way you wrote it?

Reply:
oups just read in the riot chat that you weren't interested in code feedback haha :p

Reply:
no worries, all good, thanks